src/newrelic/collector/client.go

```diff
@@ -6,310 +6,302 @@
+    if nil != err {
+        return RPMResponse{Err: err}
+    }
 
-	req.Header.Add("Accept-Encoding", "identity, deflate")
-	req.Header.Add("Content-Type", "application/octet-stream")
-	req.Header.Add("User-Agent", userAgent)
-	req.Header.Add("Content-Encoding", "deflate")
+    req.Header.Add("Accept-Encoding", "identity, deflate")
+    req.Header.Add("Content-Type", "application/octet-stream")
+    req.Header.Add("User-Agent", cs.userAgent())
+    req.Header.Add("Content-Encoding", "deflate")
 
-	for k, v := range requestHeadersMap {
+	for k, v := range cmd.RequestHeadersMap {
 		req.Header.Add(k, v)
 	}
 
-	resp, err := c.httpClient.Do(req)
-	if err != nil {
-		return nil, err
-	}
-
-	defer resp.Body.Close()
-
-	switch resp.StatusCode {
-	case 200:
-		// Nothing to do.
-	case 202:
-		// "New" style P17 success response code with no JSON payload to decode
-		// FIXME - this just gets something working but need to change
-		//         all this code to not expect a reply payload on success 
-		//         as was the case with P16
-		//         Will be replaced by proper P17 response code handling
-		return []byte(""), nil
-	case 401:
-		return nil, ErrUnauthorized
-	case 413:
-		return nil, ErrPayloadTooLarge
-	case 415:
-		return nil, ErrUnsupportedMedia
-	default:
-		// If the response code is not 200, then the collector may not return
-		// valid JSON
-		return nil, fmt.Errorf("unexpected collector HTTP status code: %d",
-			resp.StatusCode)
-	}
-
-	b, err := ioutil.ReadAll(resp.Body)
-	if nil != err {
-		return nil, err
-	}
-	return parseResponse(b)
-}
-
-func (c *clientImpl) Execute(cmd Cmd) ([]byte, error) {
-	data, err := cmd.Collectible.CollectorJSON(false)
-	if nil != err {
-		return nil, fmt.Errorf("unable to create json payload for '%s': %s",
-			cmd.Name, err)
-	}
-
-	var audit []byte
-
-	if log.Auditing() {
-		audit, err = cmd.Collectible.CollectorJSON(true)
-		if nil != err {
-			log.Errorf("unable to create audit json payload for '%s': %s", cmd.Name, err)
-			audit = data
-		}
-		if nil == audit {
-			audit = data
-		}
-	}
-
-	url := cmd.url(false)
-	cleanURL := cmd.url(true)
+    resp, err := c.httpClient.Do(req)
+    if err != nil {
+        return RPMResponse{Err: err}
+    }
 
-	log.Audit("command='%s' url='%s' payload={%s}", cmd.Name, url, audit)
-	log.Debugf("command='%s' url='%s' payload={%s}", cmd.Name, cleanURL, data)
+    defer resp.Body.Close()
 
-	resp, err := c.perform(url, data, cmd.userAgent(), cmd.RequestHeadersMap)
-	if err != nil {
-		log.Debugf("attempt to perform %s failed: %q, url=%s",
-			cmd.Name, err.Error(), cleanURL)
-	}
+    r := newRPMResponse(resp.StatusCode)
 
-	log.Audit("command='%s' url='%s', response={%s}", cmd.Name, url, resp)
-	log.Debugf("command='%s' url='%s', response={%s}", cmd.Name, cleanURL, resp)
-
-	return resp, err
+    body, err := ioutil.ReadAll(resp.Body)
+    if nil != err {
+        r.Err = err
+    } else {
+        r.Body, r.Err = parseResponse(body)
+    }
```

Comment:
```suggestion
    body, err := ioutil.ReadAll(resp.Body)
    if nil != err {
        r.Err = err
    } else {
        r.Body = body    
    }
```
Please remove call to `parseResponse` body - we no longer want to parse collector response, we want to use HTTP Status Codes. Also with P17 `body` will be empty and parseBody calls json.Unmarshal(), which expects a valid JSON string. Without it, it'll throw "unexpected end of JSON input" error.

Reply:
I could use a little help on this aspect. Originally, I had removed the parseResponse() call. However, this caused issues with our LASP tests, [as seen here on Jenkins](https://phpagent-build.pdx.vm.datanerd.us/job/php-pull-request-parallel-8-1/label=centos5-64-nrcamp/149/console). Our policy map is being sent in the body of the response. I do not know of the desired way forward from here.

Reply:
I think the parsereply still needs to stay in there so we can extract security policies when they exist.  If you create an Unmarshal for the connectreply like go-agent does, it'll catch for when the json is empty and extract security policy when it is not: https://github.com/newrelic/go-agent/blob/06c801d5571056abac8ac9dfa07cf12ca869e920/v3/internal/connect_reply.go#L254

Reply:
Digging deeper through go-agent code reveals that go-agent actually does parse the reply like daemon's `parseResponse` function: 
[newrelic/go-agent/internal/collector.go:268-276](https://github.com/newrelic/go-agent/blob/4b46fc5b40e5398695413097b5c461679a74eafd/internal/collector.go#L276)
```
	resp := CollectorRequest(call, cs)
	if nil != resp.Err {
		return nil, resp
	}

	var preconnect struct {
		Preconnect PreconnectReply `json:"return_value"`
	}
	err = json.Unmarshal(resp.body, &preconnect)

```